ci: build & push images to GAR with base image caching

cleot opened the pull request:
partially replaces #157

[latest test run](https://github.com/BitcreditProtocol/Wildcat/actions/runs/15143209427)


Adds a GitHub Actions workflow that builds a reusable base Docker image, caches it in Google Artifact Registry (GAR), and then builds and pushes multiple service images using that cache.

- Introduces a two-stage workflow: one job to build and upload the base image artifact, and another to load that base and build/push service images.
- Configures remote caching for faster rebuilds
- Adds provenance attestation for each pushed image.

repo:
<img width="709" alt="image" src="https://github.com/user-attachments/assets/f02932ee-05c5-4121-b67e-1887854da062" />


package:
<img width="941" alt="image" src="https://github.com/user-attachments/assets/c40eaed0-68c8-4bc4-9afc-b9f49445e609" />

codingpeanut157 reviewed (changes requested):
I would remove the commented line at 5 that builds on any push to branches and leave only schedule and tags with semver.

Copilot also suggests to pin the actions with the hash of the commit rather than the version.
Given these are critical services I would change them accordingly

cleot commented:
> I would remove the commented line at 5 that builds on any push to branches and leave only schedule and tags with semver.
> 
> Copilot also suggests to pin the actions with the hash of the commit rather than the version. Given these are critical services I would change them accordingly


- removed line 5 (build on any branch for testing)
- pinned all actions with commit hashes
- added provenance attestations
- changed gcp auth from service account key to workload identity federation
- removed unnecessary comments
- added registry repository cleanup policy (keep 10 recent versions, delete older than 30d, keep tagged versions)

zupzup commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,171 @@
+      attestations: write
+    strategy:
+      fail-fast: false
+      #max-parallel: 1
```
leftover?

cleot replied:
yes, will remove in next version if we keep everything parallel and still use an extra step for base image caching, will also need to set build-args to be able to pull base-image from registry and use gha cache with default buildx driver

zupzup reviewed (approved):
Nice :+1:
